File: ios/Mailrs/Accounts/SMTPSession.swift
import Foundation
import Network

/// A conversation with a submission server.
///
/// Submission, not delivery: this speaks to the provider's own server
/// on 465 or 587 with the account's credential. Sending mail from a
/// phone straight to the recipient's MX fails SPF and DMARC at every
/// receiver, so it is not an option and not a preference.
actor SMTPSession {
    enum Failure: Error, Equatable {
        case unreachable(String)
        /// The credential was refused — a button to press.
        case refused(String)
        /// The server said no. `permanent` decides whether trying
        /// again could ever work.
        case rejected(code: Int, text: String, permanent: Bool)
        case closed
    }

    private let transport: ByteTransport
    private let host: String
    private var buffer = Data()
    /// Whether TLS was there from the first byte.
    private let implicitTLS: Bool

    /// `port == 465` is TLS from the first byte; 587 is STARTTLS.
    ///
    /// Both are used in the wild and providers disagree about which
    /// they offer, so the port decides rather than a setting nobody
    /// can answer. Outlook and iCloud are 587-only and the provider
    /// table hands out that port, so a session that cannot do the
    /// second is a mailbox that can never send.
    init(host: String, port: UInt16) {
        self.host = host
        implicitTLS = port == 465
        if port == 465 {
            transport = TLSTransport(host: host, port: port)
        } else {
            transport = UpgradableTransport(host: host, port: port)
        }
    }

    /// For tests: a scripted server, and the conversation under test.
    init(host: String, port: UInt16, transport: ByteTransport) {
        self.host = host
        self.transport = transport
        implicitTLS = port == 465
    }

    func connect(helo: String) async throws {
        do {
            try await transport.connect()
        } catch let e as TransportFailure {
            throw failure(e)
        }
        _ = try await readReply()  // the greeting
        let greeted = try await command("EHLO \(helo)")
        guard !implicitTLS else { return }

        // **No downgrade.** A server that does not offer to encrypt is
        // not argued with — the credential simply does not go there. A
        // stripped capability list is what somebody in the middle
        // produces, and refusing is the only answer to it.
        guard greeted.text.uppercased().contains("STARTTLS") else {
            throw Failure.refused("the server did not offer to encrypt the connection")
        }
        let upgrade = try await command("STARTTLS")
        guard upgrade.code == 220 else {
            throw Failure.refused("the server refused to encrypt the connection")
        }
        do {
            try await transport.upgradeToTLS()
        } catch let e as TransportFailure {
            throw failure(e)
        }
        // Again, because everything the server said before the upgrade
        // was said by whoever was on the wire at the time.
        _ = try await command("EHLO \(helo)")
    }

    private func failure(_ e: TransportFailure) -> Failure {
        switch e {
        case let .unreachable(why): return .unreachable(why)
        case .closed: return .closed
        case .cannotUpgrade: return .refused("this connection could not be encrypted")
        }
    }

    func close() async { await transport.close() }

    /// Sign in with a password or an access token.
    ///
    /// An access token is not a password: it goes through XOAUTH2, and
    /// sending one through `AUTH PLAIN` is refused — after which the
    /// person is told their password is wrong for an account whose
    /// credentials are fine.
    func authenticate(user: String, secret: String, oauth: Bool) async throws {
        var reply: SMTP.Reply
        if oauth {
            reply = try await command("AUTH XOAUTH2 \(SMTP.authXOAuth2(user: user, token: secret))")
        } else {
            reply = try await command("AUTH PLAIN \(SMTP.authPlain(user: user, password: secret))")
        }
        // **334 first.** It is inside the range `isPositive` calls
        // success, so testing that first made this branch unreachable
        // and reported every refused token as signed in — which is
        // exactly what the note below says must not happen. A provider
        // only sends 334 when a token is genuinely bad, so a real
        // server never shows it; a scripted one does.
        //
        // A provider that rejects an OAuth token answers 334 with a
        // base64 error rather than a final code, and waits for an
        // empty line before sending one.
        if reply.code == 334 {
            let final = try await command("")
            throw Failure.refused(final.text)
        }
        if reply.isPositive { return }
        if SMTP.isAuthenticationFailure(code: reply.code, text: reply.text) {
            throw Failure.refused(reply.text)
        }
        throw Failure.rejected(code: reply.code, text: reply.text, permanent: reply.isPermanent)
    }

    /// Hand one message over.
    func send(from: String, to: [String], message: String) async throws {
        try await send(from: from, to: to, message: AnySequence([message]))
    }

    /// Hand one message over, in as many pieces as it comes in.
    ///
    /// **Streamed rather than assembled.** A 25 MB attachment built
    /// into one string and dot-stuffed into another is several times
    /// its own size in memory at once, and on a phone that is a
    /// process the system kills — which looks exactly like mail that
    /// vanished. Here each piece is stuffed and written as it arrives,
    /// so what is held is one piece.
    ///
    /// The stuffing is `DotStuffer`'s, not `SMTP.dotStuffed`'s,
    /// because a piece can end on the line break whose next line
    /// begins with a dot — and a stuffer with no memory of that
    /// truncates the message there while it still arrives looking
    /// complete.
    func send(from: String, to: [String], message: AnySequence<String>) async throws {
        try await expect(try await command("MAIL FROM:<\(from)>"))
        for rcpt in to {
            try await expect(try await command("RCPT TO:<\(rcpt)>"))
        }
        let start = try await command("DATA")
        guard start.code == 354 else {
            throw Failure.rejected(
                code: start.code, text: start.text, permanent: start.isPermanent)
        }
        let stuffer = DotStuffer()
        var last = ""
        for piece in message where !piece.isEmpty {
            try await write(stuffer.feed(piece))
            last = piece
        }
        // The terminator needs a line of its own, and a message that
        // already ended on one must not gain a blank line — some
        // servers keep it and the reader sees it.
        try await write(last.hasSuffix("\r\n") ? ".\r\n" : "\r\n.\r\n")
        try await expect(try await readReply())
        _ = try? await command("QUIT")
    }

    // MARK: - the wire

    private func expect(_ reply: SMTP.Reply) throws {
        guard reply.isPositive else {
            throw Failure.rejected(
                code: reply.code, text: reply.text, permanent: reply.isPermanent)
        }
    }

    private func command(_ text: String) async throws -> SMTP.Reply {
        try await write(text + "\r\n")
        return try await readReply()
    }

    /// A reply, however many lines it takes.
    ///
    /// `250-STARTTLS` continues and `250 OK` ends; reading only the
    /// first line leaves the rest in the buffer, and the next command
    /// then reads somebody else's answer.
    private func readReply() async throws -> SMTP.Reply {
        // **Every line, joined.** A multi-line reply is one reply, and
        // keeping only the last discards the EHLO capability list —
        // the only place `STARTTLS` is announced, and so the only way
        // to know whether an upgrade is possible at all.
        var gathered: [String] = []
        while true {
            let line = try await readLine()
            guard let r = SMTP.reply(line) else { continue }
            gathered.append(r.text)
            if !r.more {
                return SMTP.Reply(
                    code: r.code, text: gathered.joined(separator: "\n"), more: false)
            }
        }
    }

    private func write(_ text: String) async throws {
        do {
            try await transport.send(Data(text.utf8))
        } catch let e as TransportFailure {
            throw failure(e)
        }
    }

    private func readLine() async throws -> String {
        while true {
            if let r = buffer.firstRange(of: Data("\r\n".utf8)) {
                let line = String(decoding: buffer[..<r.lowerBound], as: UTF8.self)
                buffer.removeSubrange(..<r.upperBound)
                return line
            }
            let chunk = try await receive()
            if chunk.isEmpty { throw Failure.closed }
            buffer.append(chunk)
        }
    }

    private func receive() async throws -> Data {
        do {
            return try await transport.receive()
        } catch let e as TransportFailure {
            throw failure(e)
        }
    }
}

File: android/app/src/main/kotlin/jp/golia/mailrs/accounts/SmtpSession.kt
package jp.golia.mailrs.accounts

import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.Writer
import java.net.InetSocketAddress
import java.net.Socket
import javax.net.ssl.SSLSocket
import javax.net.ssl.SSLSocketFactory
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * A conversation with a submission server.
 *
 * Submission, not delivery: this speaks to the provider's own server
 * on 465 with the account's credential. Sending mail from a phone
 * straight to the recipient's MX fails SPF and DMARC at every
 * receiver, so it is not an option and not a preference.
 */
class SmtpSession(private val host: String, private val port: Int) : AutoCloseable {
    sealed class Failure(message: String) : Exception(message) {
        class Unreachable(val why: String) : Failure(why)

        /** The credential was refused — a button to press. */
        class Refused(val detail: String) : Failure(detail)

        /**
         * The server said no. [permanent] decides whether trying again
         * could ever work.
         */
        class Rejected(val code: Int, val text: String, val permanent: Boolean) : Failure(text)

        class Closed : Failure("the server closed the connection")
    }

    /**
     * Where the bytes go.
     *
     * A seam, so the conversation above it can be tested without a
     * server. The rules this session enforces — the fourth character of
     * a reply is what says whether more is coming, a 334 is a refusal
     * and not a success, a body line beginning with a dot must be
     * stuffed — are rules about what arrives, and a real server is
     * exactly the thing that will not send the awkward case on demand.
     */
    interface Transport {
        fun readLine(): String
        fun write(text: String)
        fun close()
    }

    /** Swapped in by tests; null means the real socket. */
    internal var transport: Transport? = null

    private var socket: SSLSocket? = null
    private var reader: BufferedReader? = null
    private var writer: Writer? = null

    /**
     * Open the connection, encrypt it, and greet.
     *
     * Two ways in, because the world has two: **465 is encrypted from
     * the first byte**, and **587 starts in the clear and upgrades**.
     * Refusing the second is not caution — Outlook and iCloud are
     * 587-only, and the provider table hands out that port, so a
     * session that refuses it is a mailbox that can never send.
     *
     * **No downgrade either way.** If the upgrade does not happen the
     * connection is dropped rather than used: an unencrypted
     * submission carries the password in the clear, and the point of
     * refusing is that nobody can make it happen by standing in the
     * middle and removing STARTTLS from the capability list.
     */
    suspend fun connect(helo: String, timeoutMs: Int = 20_000) = withContext(Dispatchers.IO) {
        if (transport != null) {
            // A scripted server: the conversation is what is under
            // test, and there is no socket to upgrade. The conversation
            // is otherwise the same one, **including the second EHLO**
            // — a test path that skips a command is a test of a
            // different protocol.
            readReply()
            val greeted = command("EHLO $helo")
            if (port != 465) {
                startTls(helo, greeted)
                command("EHLO $helo")
            }
            return@withContext
        }
        if (port != 465) {
            connectWithStartTls(helo, timeoutMs)
            return@withContext
        }
        try {
            val plain = Socket()
            plain.connect(InetSocketAddress(host, port), timeoutMs)
            plain.soTimeout = 60_000
            val s = (SSLSocketFactory.getDefault() as SSLSocketFactory)
                .createSocket(plain, host, port, true) as SSLSocket
            s.sslParameters = s.sslParameters.apply { endpointIdentificationAlgorithm = "HTTPS" }
            s.startHandshake()
            socket = s
            reader = BufferedReader(InputStreamReader(s.inputStream, Charsets.UTF_8))
            writer = s.outputStream.writer(Charsets.UTF_8)
        } catch (e: Exception) {
            throw Failure.Unreachable(e.message ?: e.toString())
        }
        readReply()
        command("EHLO $helo")
    }

    /** 587 and friends: greet in the clear, upgrade, greet again. */
    private fun connectWithStartTls(helo: String, timeoutMs: Int) {
        val plain: Socket
        try {
            plain = Socket()
            plain.connect(InetSocketAddress(host, port), timeoutMs)
            plain.soTimeout = 60_000
            reader = BufferedReader(InputStreamReader(plain.inputStream, Charsets.ISO_8859_1))
            writer = plain.outputStream.writer(Charsets.UTF_8)
        } catch (e: Exception) {
            throw Failure.Unreachable(e.message ?: e.toString())
        }
        readReply()
        val greeted = command("EHLO $helo")
        startTls(helo, greeted)
        try {
            val s = (SSLSocketFactory.getDefault() as SSLSocketFactory)
                .createSocket(plain, host, port, true) as SSLSocket
            // Names are checked against the certificate. Without this
            // any certificate a machine on the path can produce is
            // accepted, which is the whole of the protection gone.
            s.sslParameters = s.sslParameters.apply {
                endpointIdentificationAlgorithm = "HTTPS"
            }
            s.startHandshake()
            socket = s
            reader = BufferedReader(InputStreamReader(s.inputStream, Charsets.ISO_8859_1))
            writer = s.outputStream.writer(Charsets.UTF_8)
        } catch (e: Exception) {
            runCatching { plain.close() }
            throw Failure.Unreachable(e.message ?: e.toString())
        }
        // Again, because everything the server said before the upgrade
        // was said by whoever was on the wire at the time.
        command("EHLO $helo")
    }

    /**
     * Ask for the upgrade, and refuse to go on without it.
     *
     * The capability list is read from the greeting rather than
     * trusted: a server that does not offer STARTTLS is not asked, and
     * one that offers it and then refuses is dropped. Neither case
     * continues in the clear.
     */
    private fun startTls(helo: String, greeted: Smtp.Reply) {
        if (!greeted.text.uppercase().contains("STARTTLS")) {
            throw Failure.Refused("the server did not offer to encrypt the connection")
        }
        val upgrade = command("STARTTLS")
        if (upgrade.code != 220) {
            throw Failure.Refused("the server refused to encrypt the connection")
        }
    }

    override fun close() {
        transport?.let {
            runCatching { it.close() }
            return
        }
        runCatching { socket?.close() }
        socket = null
    }

    /**
     * Sign in with a password or an access token.
     *
     * A provider that rejects an OAuth token answers 334 with a base64
     * error rather than a final code, and waits for an empty line
     * before sending one. Reading the 334 as success authenticates
     * every refused token.
     */
    suspend fun authenticate(user: String, secret: String, oauth: Boolean) =
        withContext(Dispatchers.IO) {
            val reply = if (oauth) {
                command("AUTH XOAUTH2 ${Smtp.authXOAuth2(user, secret)}")
            } else {
                command("AUTH PLAIN ${Smtp.authPlain(user, secret)}")
            }
            // **334 first.** It is inside the 2xx-3xx range that
            // `isPositive` calls success, so testing that first made
            // this branch unreachable and reported every refused token
            // as signed in — which is precisely what the note above
            // says must not happen. Caught by a scripted server, never
            // by a real one: a real one only sends 334 when a token is
            // genuinely bad.
            if (reply.code == 334) {
                val final = command("")
                throw Failure.Refused(final.text)
            }
            if (reply.isPositive) return@withContext
            throw if (Smtp.isAuthenticationFailure(reply.code, reply.text)) {
                Failure.Refused(reply.text)
            } else {
                Failure.Rejected(reply.code, reply.text, reply.isPermanent)
            }
        }

    /** Hand one message over. */
    suspend fun send(from: String, to: List<String>, message: String) =
        send(from, to, sequenceOf(message))

    /**
     * Hand one message over, in as many pieces as it comes in.
     *
     * **Streamed rather than assembled.** A 25 MB attachment built into
     * one string and dot-stuffed into another is several times its own
     * size in memory at once, and on a phone that is a process the
     * system kills — which looks exactly like mail that vanished. Here
     * each piece is stuffed and written as it arrives, so what is held
     * is one piece.
     *
     * The stuffing is [DotStuffer]'s, not [Smtp.dotStuffed]'s, because
     * a piece can end on the line break whose next line begins with a
     * dot — and a stuffer with no memory of that truncates the message
     * there while it still arrives looking complete.
     */
    suspend fun send(from: String, to: List<String>, message: Sequence<String>) =
        withContext(Dispatchers.IO) {
            expect(command("MAIL FROM:<$from>"))
            for (rcpt in to) expect(command("RCPT TO:<$rcpt>"))
            val start = command("DATA")
            if (start.code != 354) {
                throw Failure.Rejected(start.code, start.text, start.isPermanent)
            }
            val stuffer = DotStuffer()
            var last = ""
            for (piece in message) {
                if (piece.isEmpty()) continue
                write(stuffer.feed(piece))
                last = piece
            }
            // The terminator needs a line of its own, and a message
            // that already ended on one must not gain a blank line —
            // some servers keep it and the reader sees it.
            val terminator = when {
                last.endsWith("\r\n") -> ".\r\n"
                else -> "\r\n.\r\n"
            }
            write(terminator)
            expect(readReply())
            runCatching { command("QUIT") }
            Unit
        }

    // MARK: the wire

    private fun expect(reply: Smtp.Reply) {
        if (!reply.isPositive) {
            throw Failure.Rejected(reply.code, reply.text, reply.isPermanent)
        }
    }

    private fun command(text: String): Smtp.Reply {
        write(text + "\r\n")
        return readReply()
    }

    /**
     * A reply, however many lines it takes.
     *
     * `250-STARTTLS` continues and `250 OK` ends; reading only the
     * first line leaves the rest in the buffer, and the next command
     * then reads somebody else's answer.
     */
    private fun readReply(): Smtp.Reply {
        // **Every line, joined.** A multi-line reply is one reply, and
        // throwing away all but the last discards the EHLO capability
        // list — which is the only place `STARTTLS` is announced, and
        // the only way to know whether an upgrade is even possible. It
        // is also where a server explains a refusal, in the lines
        // before the one carrying the code.
        val gathered = StringBuilder()
        while (true) {
            val line = readLine()
            val r = Smtp.reply(line) ?: continue
            if (gathered.isNotEmpty()) gathered.append('\n')
            gathered.append(r.text)
            if (!r.more) return r.copy(text = gathered.toString())
        }
    }

    private fun write(text: String) {
        transport?.let {
            it.write(text)
            return
        }
        val w = writer ?: throw Failure.Closed()
        w.write(text)
        w.flush()
    }

    private fun readLine(): String {
        transport?.let { return it.readLine() }
        return reader?.readLine() ?: throw Failure.Closed()
    }
}
